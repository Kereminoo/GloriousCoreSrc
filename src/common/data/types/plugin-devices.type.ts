import { PluginDeviceRecord } from "../records/plugin-device.record";

export type PluginDevices = 
{
    Mouse: PluginDeviceRecord[];
    Keyboard: PluginDeviceRecord[];
    valueE: PluginDeviceRecord[];
    MouseDock: PluginDeviceRecord[];
}