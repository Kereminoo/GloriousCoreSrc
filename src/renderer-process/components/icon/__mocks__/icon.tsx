import React from 'react';
import { IconProps } from '../icon';

const IconMock: React.FC<IconProps> = ({ type, color, size }) => (
  <div>
    {size} sized {type} icon in {color} color
  </div>
);

export default IconMock;
