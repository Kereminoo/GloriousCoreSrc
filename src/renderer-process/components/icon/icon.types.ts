import { Color } from '../component.types';

export enum IconSize {
  XSmall = 'xsmall',
  Smaller = 'smaller',
  Small = 'small',
  Medium = 'medium',
  Large = 'large',
  Larger = 'larger',
  ExtraLarge = 'extralarge',
}
export enum UniqueIconSize {
  ModelO = 'model-o',
  ModelD = 'model-d',
  ModelI = 'model-i',
  valueJ = 'valueJ',
  GMMKPRO = 'gmmk-pro',
  GMMK265 = 'gmmk2-65',
  GMMK296 = 'gmmk2-96',
  Keyboard100 = 'keyboard-100',
  Numpad = 'numpad',
  Drag = 'drag',
  LayerTop = 'layer-top',
  LayerUnderneath = 'layer-underneath',
  NavigationArrow = 'navigation-arrow',
  DocumentMagnifyingGlass = 'document-magnifying-glass',
  USBSymbol = 'usb-symbol',
}
export const UniqueIconSizeMap  = new Map([
  [ "model-o", { width: "26px", height: "40px" } ],
  [ "model-d", { width: "28px", height: "40px" } ],
  [ "model-i", { width: "30px", height: "40px" } ],
  [ "valueJ", { width: "26px", height: "40px" } ],
  [ "gmmk-pro", { width: "71px", height: "36px" } ],
  [ "gmmk2-65", { width: "74px", height: "32px" } ],
  [ "gmmk2-96", { width: "81px", height: "36px" } ],
  [ "numpad", { width: "32px", height: "32px" } ],
  [ "drag", { width: "12px", height: "20px" } ],
  [ "layer-top", { width: "47px", height: "25px" } ],
  [ "layer-underneath", { width: "47px", height: "17px" } ],
  [ "navigation-arrow", { width: "19px", height: "14px" } ],
  [ "document-magnifying-glass", { width: "28px", height: "35px" } ],
  [ "usb-symbol", { width: "12px", height: "16px" } ],
]);

export const iconColorsOptions: (Color | undefined)[] = [
  Color.Base100,
  Color.Base60,
  Color.Base70,
  Color.Base50,
  Color.Base30,
  Color.Base20,
  Color.Base20,
  Color.Glorange60,
  Color.Glorange50,
  Color.Glorange20,
  undefined,
];

export type IconColor =
  | Color.Base100
  | Color.Base60
  | Color.Base70
  | Color.Base50
  | Color.Base30
  | Color.Base20
  | Color.Glorange60
  | Color.Glorange50
  | Color.Glorange20
  | Color.GreenDark60
  | Color.GreenDark40
  | Color.RedDark60
  | Color.GIDSuccessDark
  | Color.GIDSuccessLight
  | Color.GIDUploadDark
  | Color.GIDUploadLight
  | Color.GIDSyncingDark
  | Color.GIDSyncingLight
  | Color.GIDDisconnectDark
  | Color.GIDDisconnectLight
  | Color.GIDLogoMark;

export enum IconType {

  // Status
  SuccessCheck = 'success-check',
  FailCross = 'fail-cross',

  // Utility
  SaveCheck = 'save-check',
  CancelCross = 'cancel-cross',
  InformationOutline = 'information-outline',
  Trash = 'trash',
  ExclamationPoint = 'exclamation-point',
  QuestionMark = 'question-mark',
  NavigationArrow = 'navigation-arrow',
  CircleArrow = 'circle-arrow',
  UpArrowOutline = 'up-arrow-outline',
  DocumentMagnifyingGlass = 'document-magnifying-glass',
  Hourglass = 'hourglass',
  USBSymbol = 'usb-symbol',
  VerticalEllipses = 'vertical-ellipses',
  Drag = 'drag',
  Plus = 'plus',
  PlusOutline = 'plus-outline',
  Edit = 'edit',
  FloppyDisk = 'floppy-disk',
  Undo = 'undo',
  SaveCheckRounded = 'save-check-rounded',


  // Navigation
  GloriousLogo = 'glorious-logo',
  CogFilled = 'cog-filled',
  RGBSync = 'rgb-sync',
  ModelODevice = 'model-o-device',
  ModelDDevice = 'model-d-device',
  ModelIDevice = 'model-i-device',
  GMMK265Device = 'gmmk2-65-device',
  GMMK296Device = 'gmmk2-96-device',
  GMMKPRODevice = 'gmmk-pro-device',
  NumpadDevice = 'numpad-device',
 
  // Connected Devices
  BatteryLow = 'battery-low',
  BatteryMedium = 'battery-medium',
  BatteryHigh = 'battery-high',
  WirelessConnectionStrong = 'wireless-connection-strong',
  WirelessConnectionMedium = 'wireless-connection-medium',
  WirelessConnectionWeak = 'wireless-connection-weak',
  WirelessConnectionDisconnected = 'wireless-connection-disconnected',
 
 // Settings
  ProfilePicture = 'profile-picture',
  Sun = 'sun',
  Moon = 'moon',
 
  // GloriousID
  GloriousID = 'glorious-id',
  CloudCheck = 'cloud-check',
  CloudUpload = 'cloud-upload',
  CloudSyncing = 'cloud-syncing',
  CloudDisconnected = 'cloud-disconnected',
  WireframeGlobe = 'wireframe-globe',
  Shop = 'shop',
  Filter = 'filter',
  Sort = 'sort',

  // Pairing

  // Devices
  Home = 'home',
  CogOutline = 'cog-outline',
  Lightbulb = 'lightbulb',
  LightningBolt = 'lightning-bolt',
  Speedometer = 'speedometer',
  Keybinding = 'keybinding',
  Actuation = 'actuation',
  AdvancedKeys = 'advanced-keys',
  DynamicKeystroke = 'dynamic-keystroke',
  ModTap = 'mod-tap',
  ToggleKey = 'toggle-key',
  LayerTop = 'layer-top',
  LayerUnderneath = 'layer-underneath',

  // Macros
  MacroNoRepeat = 'macro-no-repeat',
  MacroRepeatWhileHolding = 'macro-repeat-while-holding',
  MacroToggle = 'macro-toggle',

  // these icons were "modern ui" icons; not custom glorious icons
  // Alert = 'alert',
  // ArrowLeft = 'arrow-left',
  // BatteryCharging = 'battery-charging',
  // ChatProcessing = 'chat-processing',
  // Close = 'close',
  // Copy = 'copy',
  // Delete = 'delete',
  // GamepadSquare = 'gamepad-square',
  // valueEMic = 'valueE-mic',
  // Mic = 'mic',
  // MicOff = 'mic-off',
  // Nc = 'nc',
  // Open = 'open',
  // Pause = 'pause',
  // Play = 'play',
  // Plus = 'plus',
  // RotateLeft = 'rotate-left',
  // RotateRight = 'rotate-right',
  // Screen = 'screen',
  // SpinnerDark = 'spinner-dark',
  // SpinnerLight = 'spinner-light',
  // Stop = 'stop',
  // Tick = 'tick',
  // TransparencyMode = 'transparency-mode',
  // TransparencyOff = 'transparency-off',
  // TuneVariant = 'tune-variant',
  // Upload = 'upload',
  // VolumeUp = 'volume-up',
}
