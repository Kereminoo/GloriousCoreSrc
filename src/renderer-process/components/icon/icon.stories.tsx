import { Meta, StoryObj } from '@storybook/react';
import Icon from './icon';
import { IconSize, IconType, iconColorsOptions } from './icon.types';
import { Color } from '../component.types';

const argTypes = {
  type: {
    control: {
      type: 'select',
    },
    options: Object.values(IconType),
    description: 'The type of icon to display.',
  },
  color: {
    control: {
      type: 'select',
    },
    options: iconColorsOptions,
    description: 'Optional color setting of the icon.',
    defaultValue: {
      summary: 'Color.Base50',
    },
  },
  size: {
    control: {
      type: 'select',
    },
    options: Object.values(IconSize),
    description: 'Optional size setting of the icon.',
    defaultValue: {
      summary: 'IconSize.Small',
    },
  },
};

const meta: Meta<typeof Icon> = {
  title: 'Components/Icon',
  component: Icon,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes,
};

export default meta;

type Story = StoryObj<typeof Icon>;

export const Default: Story = {
  args: {
    type: IconType.valueEMic,
    color: Color.Base20,
    size: IconSize.Medium,
  },
};

interface IconListProps {
  size?: IconSize;
}

const IconList: React.FC<IconListProps> = ({ size }) => {
  return (
    <div style={{ display: 'flex', flexWrap: 'wrap', gap: '20px', justifyContent: 'center', alignItems: 'center' }}>
      {Object.values(IconType).map(type => (
        <div key={type} style={{ display: 'flex', flexDirection: 'column', alignItems: 'center', textAlign: 'center' }}>
          <Icon type={type} size={size} color={Color.Base20} />
          <span style={{ marginTop: '10px' }}>{type}</span>
        </div>
      ))}
    </div>
  );
};

export const All: StoryObj<typeof Icon> = {
  render: args => <IconList {...args} />,
};

All.args = {
  size: IconSize.Medium,
};
