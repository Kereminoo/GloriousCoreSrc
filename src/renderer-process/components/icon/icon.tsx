import React, { useEffect, useState } from 'react';
import { IconColor, IconSize, IconType } from './icon.types';
import { Color } from '../component.types';

export interface IconProps {
  type: IconType;
  color?: IconColor;
  size?: IconSize;
  width?: string;
  height?: string;
  secondaryColor?: IconColor;
  hoverColor?: IconColor;
  className?: string;
}

const sizeToVariable = {
  [IconSize.XSmall]: '12px',
  [IconSize.Smaller]: '16px',
  [IconSize.Small]: '18px',
  [IconSize.Medium]: '24px',
  [IconSize.Large]: '32px',
  [IconSize.Larger]: '40px',
  [IconSize.ExtraLarge]: '64px',
};

export function scaleUniqueDimensionsToSize(size: IconSize)
{
  // todo;
}

const Icon: React.FC<IconProps> = ({ type, color = Color.Base50, size = IconSize.Small, width, height, secondaryColor, hoverColor, className }) => {
  const [IconComponent, setIconComponent] = useState<React.ElementType | null>(null);

  useEffect(() => {
    const importIconComponent = async () => {
      try {
        const { default: ImportedIcon } = await import(`./icon-assets/${type}.tsx`);
        setIconComponent(() => ImportedIcon);
      } catch (error) {
        console.error(`Failed to load icon component: ${type}`, error);
        setIconComponent(null);
      }
    };

    importIconComponent();
  }, [type]);

  if (!IconComponent) return null;

  return <IconComponent 
  color={color} 
  size={(width == undefined && height == undefined) ? sizeToVariable[size] : undefined} 
  width={width} 
  height={height}
  hoverColor={hoverColor}
  secondaryColor={secondaryColor}
  className={className} />;
};

export default Icon;
