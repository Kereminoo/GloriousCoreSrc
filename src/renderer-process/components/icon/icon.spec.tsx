// @ts-ignore: Unused variable
import React from 'react';
import { render } from '@testing-library/react';
import Icon from './icon';
import { IconSize, IconType } from './icon.types';
import { act } from 'react-dom/test-utils';
import { Color } from '../component.types';

describe('Icon', () => {
  it('renders correctly with default size and color', async () => {
    let asFragment;
    await act(async () => {
      const result = render(<Icon type={IconType.BatteryMedium} />);
      asFragment = result.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with custom size', async () => {
    let asFragment;
    await act(async () => {
      const result = render(<Icon type={IconType.BatteryMedium} size={IconSize.Small} />);
      asFragment = result.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with custom color', async () => {
    let asFragment;
    await act(async () => {
      const result = render(<Icon type={IconType.BatteryMedium} color={Color.Base20} />);
      asFragment = result.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
