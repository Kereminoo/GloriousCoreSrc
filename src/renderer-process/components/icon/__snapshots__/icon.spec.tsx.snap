// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders correctly with custom color 1`] = `
<DocumentFragment>
  <svg
    fill="none"
    height="18px"
    viewBox="0 0 24 24"
    width="18px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.00001 16L4.00001 8L18 8L18 16M20 16.67L20 15L22 15L22 9L20 9L20 7.33C20 6.6 19.4 6 18.67 6L3.33001 6C2.60001 6 2.00001 6.6 2.00001 7.33L2.00001 16.67C2.00001 17.41 2.59001 18 3.33001 18L18.67 18C19.4 18 20 17.4 20 16.67ZM8 15L8 9L5.00001 9L5.00001 15L8 15ZM12.5 15L12.5 9L9.5 9L9.5 15L12.5 15Z"
      fill="var(--base-shades-20)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon renders correctly with custom size 1`] = `
<DocumentFragment>
  <svg
    fill="none"
    height="18px"
    viewBox="0 0 24 24"
    width="18px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.00001 16L4.00001 8L18 8L18 16M20 16.67L20 15L22 15L22 9L20 9L20 7.33C20 6.6 19.4 6 18.67 6L3.33001 6C2.60001 6 2.00001 6.6 2.00001 7.33L2.00001 16.67C2.00001 17.41 2.59001 18 3.33001 18L18.67 18C19.4 18 20 17.4 20 16.67ZM8 15L8 9L5.00001 9L5.00001 15L8 15ZM12.5 15L12.5 9L9.5 9L9.5 15L12.5 15Z"
      fill="var(--base-shades-50)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Icon renders correctly with default size and color 1`] = `
<DocumentFragment>
  <svg
    fill="none"
    height="18px"
    viewBox="0 0 24 24"
    width="18px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4.00001 16L4.00001 8L18 8L18 16M20 16.67L20 15L22 15L22 9L20 9L20 7.33C20 6.6 19.4 6 18.67 6L3.33001 6C2.60001 6 2.00001 6.6 2.00001 7.33L2.00001 16.67C2.00001 17.41 2.59001 18 3.33001 18L18.67 18C19.4 18 20 17.4 20 16.67ZM8 15L8 9L5.00001 9L5.00001 15L8 15ZM12.5 15L12.5 9L9.5 9L9.5 15L12.5 15Z"
      fill="var(--base-shades-50)"
    />
  </svg>
</DocumentFragment>
`;
